packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-granules.js

```diff
@@ -574,3 +584,85 @@ test('writeGranules() does not persist records to Dynamo or Postgres if Postgres
     await t.context.granulePgModel.exists(knex, { granule_id: granuleId })
   );
 });
+
+test.serial('writeGranules() writes a granule and marks as failed if any file writes fail', async (t) => {
+  const {
+    cumulusMessage,
+    knex,
+    collectionCumulusId,
+    executionCumulusId,
+    providerCumulusId,
+    granuleModel,
+    granuleId,
+  } = t.context;
+
+  cumulusMessage.meta.status = 'completed';
+
+  cumulusMessage.payload.granules[0].files[0].bucket = undefined;
+  cumulusMessage.payload.granules[0].files[0].key = undefined;
+
+  await writeGranules({
+    cumulusMessage,
+    collectionCumulusId,
+    executionCumulusId,
+    providerCumulusId,
+    knex,
+    granuleModel,
+  });
+
+  t.true(await granuleModel.exists({ granuleId }));
+
+  const pgGranule = await t.context.granulePgModel.get(knex, { granule_id: granuleId });
+
+  t.is(pgGranule.status, 'failed');
+  t.deepEqual(pgGranule.error.Error, 'Failed writing files to Postgres.');
+});
+
+test.serial('writeGranules() writes all valid files if any non-valid file fails', async (t) => {
```

Comment:
I think that this would be better as a unit test of `_writeFiles` directly. Since that is where the relevant logic under test here lives

Reply:
Yeah so I'm going back-and-forth on how to structure these tests and which functions should be private v. public. What I like about testing the logic from `writeGranules` is that we're also getting the `generateFilePgRecords` logic under test here as well. That would previously throw errors for non-valid files and has been updated.

The tests for this file are currently set up to only test the exported functions, which I think is cool. I could make this change and export `_writeFiles` (minus the underscore) but I also like limiting our exports here.

Thoughts? I don't feel super strongly. Just strongly enough to want to discuss first 😄

Reply:
I don't think I would worry about whether we have to export an extra function to test it or not. 

I see what you mean `writeGranules` testing the logic for generating and writing the records, so that's fine. Though it could equally be achieved by unit tests for both of those functions. But that gets into unit testing philosophy and it's not that important here